Fix getendpoints command for compound keys containing ':'

tarasveretilnyk opened the pull request:
Before, the `nodetool getendpoints` expected the key as one string separated by : (for example 1:val:ue). This caused errors if any part of the key had a colon because it was unclear whether a colon was a separator or part of the key.

This change adds a new API endpoint, `/storage_service/natural_endpoints/v2/{keyspace}`, which accepts composite partition keys as multiple key_component query parameters (e.g., ?key_component=1&key_component=val:ue). The `nodetool getendpoints` command was updated to support a new `--key-components` option, allowing users to pass key components as an array. The client and test infrastructure were extended to support multiple values for a query parameter, and tests were added to verify correct behavior with composite keys.

The previous method of passing partition keys as colon-separated strings is preserved for backward compatibility.

Backport is not required, since this change relies on recent Seastar updates

Fixes #16596

mykaul commented:
> Backport is not required

Not required, but why wouldn't we backport such a fix?

tarasveretilnyk commented:
> > Backport is not required
> 
> Not required, but why wouldn't we backport such a fix?

Because this fix requires recent seastar patch(https://github.com/scylladb/seastar/pull/2872/commits/46b66ec615e194b9b022cbd53b08e1e00e8f908c)

denesb commented:
> > > Backport is not required
> > 
> > 
> > Not required, but why wouldn't we backport such a fix?
> 
> Because this fix requires recent seastar patch([scylladb/seastar@46b66ec](https://github.com/scylladb/seastar/commit/46b66ec615e194b9b022cbd53b08e1e00e8f908c))

Requiring a seastar patch on its own is not automatic exclusion for backport. We have done it before. We use the same criteria for seastar.git patch backport that we use for scylladb.git ones: what is the risk/benefit ratio? If it is reasonable, we can backport it.

tarasveretilnyk commented:
> > > > Backport is not required
> > > 
> > > 
> > > Not required, but why wouldn't we backport such a fix?
> > 
> > 
> > Because this fix requires recent seastar patch([scylladb/seastar@46b66ec](https://github.com/scylladb/seastar/commit/46b66ec615e194b9b022cbd53b08e1e00e8f908c))
> 
> Requiring a seastar patch on its own is not automatic exclusion for backport. We have done it before. We use the same criteria for seastar.git patch backport that we use for scylladb.git ones: what is the risk/benefit ratio? If it is reasonable, we can backport it.

I'm not sure whether it's worth backporting or not - it depends on how frequently users encounter these issues. However, the fix for non-composite partition keys has already been merged (#24829) and backported, so that probably fixes most of the problems.
Perhaps @mykaul might have more insight.

tarasveretilnyk commented on tools/scylla-nodetool.cc:
```diff
@@ -4033,7 +4033,7 @@ For more information, see: {}"
                     typed_option<sstring>("keyspace", "The keyspace to query", 1),
                     typed_option<sstring>("table", "The table to query", 1),
-                    typed_option<sstring>("key", "The partition key for which we need to find the endpoint", 1),
+                    typed_option<std::vector<sstring>>("key", "List of components of the key for which we need to find the endpoint", -1),
```
Fixed in v2

tarasveretilnyk commented on keys/keys.hh:
```diff
@@ -704,6 +704,8 @@ public:
      */
     static partition_key from_nodetool_style_string(const schema_ptr s, const sstring& key);
 
+    static partition_key from_vector(const schema_ptr s, const std::vector<sstring>& components);
```
Fixed in v2. I'm not sure whether it worth adding `nodetool` to the prefix. It accept just list of components. For me, it doesn't seem to be related to specific nodetool format.

tarasveretilnyk commented on api/api-doc/storage_service.json:
```diff
@@ -584,10 +584,10 @@
-                     "allowMultiple":false,
-                     "type":"string",
+                     "allowMultiple":true,
+                     "type":"array",
```
I also believe the new endpoint is better. Adding another query parameter to specify the key by components would have required making both parameters optional and handling them manually in the API handler, which could introduce additional confusion for users. Resolved in v2.

denesb commented on tools/scylla-nodetool.cc:
```diff
@@ -4034,6 +4049,7 @@ For more information, see: {}"
                     typed_option<sstring>("keyspace", "The keyspace to query", 1),
                     typed_option<sstring>("table", "The table to query", 1),
                     typed_option<sstring>("key", "The partition key for which we need to find the endpoint", 1),
+                    typed_option<std::vector<sstring>>("key-components", "List of components of the key for which we need to find the endpoint", -1),
```
Is this how this looks like when invoked?
```
nodetool getendpoints --key-components comp1 --key-components comp2`
```
If so, maybe better to use singular (`--key-component`).

tarasveretilnyk replied:
You can specify like you wrote. There also exists an alternative way:
`--key-components={"part1", "part2"}`

denesb commented on api/api-doc/storage_service.json:
```diff
@@ -594,6 +594,50 @@
          ]
       },
+      {
+         "path": "/storage_service/natural_endpoints/v2/{keyspace}",
```
Not sure about this name, but don't have any better idea myself.

denesb reviewed (commented):
Please also add a test to `test/nodetool`.

tarasveretilnyk commented:
v5: Added test for new nodetool param
Added support of multiple query parameters to rest_api_mock
Avoid moving params when setting request_query_params in scylla_rest_client::do_request, as they are still needed for logging and request struct is also moved later.

denesb reviewed (approved):
Looks good, I will queue this, but please follow-up with another PR which improves the documentation, we need more detail on the new param, than just "alternative to key". We also need examples on how to use this new parameter.
